docs: create generator template tutorial

Florence-Njeri opened the pull request:
Description
Tutorial to create a user-defined template

Related issue(s)
Fixes #943

Florence-Njeri commented:
@derberg @thulieblack I have put up a PR for the tutorial

thulieblack commented:
Awesome, thanks Florence

derberg commented:
yo, is this ready for final review?

derberg commented:
@pratik2315 can you have a look 🙏🏼 since it is based on my article I have a feeling better if you review first

Florence-Njeri commented:
> yo, is this ready for final review?

It is

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+
+### Add parameters to the configuration file
+
+In programming we often have different runtime environments e.g development and production. You will use different servers to spin both of these instances. In your case, you'll probably have two broker versions one for use in production and the other in development. You have defined a dev server in the AsyncAPI document:
```
```suggestion
We often have different runtime environments in programming, e.g., development and production. You will use different servers to spin both of these instances. You'll probably have two broker versions, one for production and the other for development. You have defined a dev server in the AsyncAPI document:
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+This will allow you to also define the broker you will use in production in the servers section above.
+Therefore, we can template the code `mqttBroker = 'test.mosquitto.org'` in **index.js** so the value is populated dynamically at runtime depending on the specified server environment.
+
+Generator has a **parameters** object used to define parameters you use to dynamically to modify your template code at runtime. It also supports the **server** parameter that defines the server configuration value. Navigate to **package.json** and add the snippet below:
```
```suggestion
The generator has a **parameters** object used to define parameters you use to dynamically modify your template code at runtime. It also supports the **server** parameter that defines the server configuration value. Navigate to **package.json** and add the snippet below:
```

thulieblack commented on docs/generator_template.md:
````diff
@@ -0,0 +1,568 @@
+Generator Error: This template requires the following missing params: server.
+```
+
+Update your `test:generate` script in `package.json` to inlcude the server param `test:generate": "asyncapi generate fromTemplate test/fixtures/asyncapi.yml ./ --output test/project --force-write --param server=dev "`
````
```suggestion
Update your `test:generate` script in `package.json` to include the server param `test:generate": "asyncapi generate fromTemplate test/fixtures/asyncapi.yml ./ --output test/project --force-write --param server=dev "`
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+Update your `test:generate` script in `package.json` to inlcude the server param `test:generate": "asyncapi generate fromTemplate test/fixtures/asyncapi.yml ./ --output test/project --force-write --param server=dev "`
+You can now replace the static broker from `mqttBroker = 'test.mosquitto.org'` to `mqttBroker = "${asyncapi.servers().get(params.server).url()}"` in **index.js**.
+
+Now the template code looks like:
```
```suggestion
Now the template code looks like this:
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+
+### Templating index.js with React
+
+Python takes indentation very seriously and out generated output will be python code. We therefpre need to make sure the indentation in index.js looks right so the generated code is indented properly. After templating the code in index.js, it will look like the following code snippet:
```
```suggestion
Python takes indentation very seriously, and our generated output will be Python code. We, therefore, need to make sure the indentation in index.js looks right so the generated code is indented correctly. After templating the code in index.js, it will look like the following code snippet:
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+4. Dynamically get the class name **TemperatureServiceClient** from the AsyncAPI document from the **info** object using the Parser API using the code: `asyncapi.info().title()` . It will return `Temperature Service`, then remove the spaces and add Client as a suffix.
+5. There is no templating needed in the `__init__` function, there is only hardcoded information.
+
+> If you're on the fence about which templating engine you should use in your template, check out the [react render engine](https://www.asyncapi.com/docs/tools/generator/react-render-engine) and [nunjucks render engine](https://www.asyncapi.com/docs/tools/generator/nunjucks-render-engine) documentation.
```
```suggestion
> If you're on the fence about which templating engine you should use in your template, check out the [React render engine](https://www.asyncapi.com/docs/tools/generator/react-render-engine) and [nunjucks render engine](https://www.asyncapi.com/docs/tools/generator/nunjucks-render-engine) documentation.
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,568 @@
+
+## Conclusion
+
+In this tutorial, we have created a simple AsyncAPI generator template that uses a Python MQTT client. We have explained how to use an AsyncAPI file, create an python MQTT template and use them to generate code from it and use the Paho-MQTT library in Python to connect to an MQTT broker and publish messages. We hope this tutorial has helped you understand how to create a user-defined template.
```
```suggestion
In this tutorial, we have created a simple AsyncAPI generator template that uses a Python MQTT client. We have explained how to use an AsyncAPI file, create a Python MQTT template and use them to generate code from it, and use the Paho-MQTT library in Python to connect to an MQTT broker and publish messages. We hope this tutorial has helped you understand how to create a user-defined template.
```

thulieblack commented:
Hey, @florence the PR looks great, I have additional things to add:

1. while running the `npm test` there was an error that required `rimraf` package. I think we should add that to the tutorial as well.
2. The Python loop is continual it doesn't stop running. so I suggest we add a counter to stop the `while` statement after certain intervals for instance 

```  python
counter = 0
while counter < 10:
    randomId = randrange(min_value, max_value + 1)
    client.sendTemperatureDrop(randomId)
    print("Temperature drop detected " + str(randomId) + " sent to temperature/dropped")
    client.sendTemperatureRise(randomId)
    print("Temperature rise detected " + str(randomId) + " sent to temperature/risen")
    time.sleep(1)
    counter += 1
```

This loop stops after 10 intervals what do you think?

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+## Background context
+
+There is a list of [community maintained templates](https://www.asyncapi.com/docs/tools/generator/template#generator-templates-list), but what if you require customized output from the generator? In that case, you'll create a user-defined template that generates custom output from the generator.
```
```suggestion
There is a list of [community maintained templates](https://www.asyncapi.com/docs/tools/generator/template#generator-templates-list), but what if you do not find what you need? In that case, you'll create a user-defined template that generates custom output from the generator.
```

safer to write there, because in many cases templates can be configured and generate different outputs basing on passes `parameters`

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+          type: string
+```
+
+Note:
````
I think we have this `note` component in the website that you could use here

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+Note:
+
+- To generate code, use the [AsyncAPI CLI](https://www.asyncapi.com/tools/cli). If you don't have the CLI installed, follow [this guide](https://www.asyncapi.com/docs/tools/generator/installation-guide#asyncapi-cli).
```
since this document is rendered in website, and people should read it in the website, you can do relative website links

```suggestion
- To generate code, use the [AsyncAPI CLI](https://www.asyncapi.com/tools/cli). If you don't have the CLI installed, follow [CLI installation guide](/docs/tools/cli/installation).
```

- better point to CLI docs
- better links if they are descriptive for people with accessibility problems, that use screen readers

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+Note:
+
+- To generate code, use the [AsyncAPI CLI](https://www.asyncapi.com/tools/cli). If you don't have the CLI installed, follow [this guide](https://www.asyncapi.com/docs/tools/generator/installation-guide#asyncapi-cli).
+- If you are new to AsyncAPI, check out the following docs: [template development](https://www.asyncapi.com/docs/tools/generator/template-development), which explains the minimum requirements for a template and possible features.
```
```suggestion
- If you are new to AsyncAPI Generator, check out the following docs: [template development](/docs/tools/generator/template-development), which explains the minimum requirements for a template and possible features.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+- If you are new to AsyncAPI, check out the following docs: [template development](https://www.asyncapi.com/docs/tools/generator/template-development), which explains the minimum requirements for a template and possible features.
+- You can [learn more about asynchronous messaging using MQTT](https://medium.com/python-point/mqtt-basics-with-python-examples-7c758e605d4).
+
+## Steps followed to create a template
```
`Steps followed` - not sure about grammatical correctness, what do you mean? like "steps someone should now follow" or you describe like a prerequisite steps? setup steps?

derberg replied:
btw, I like this idea of a kind of short summary of steps someone has to follow, nice overview

so maybe section name should be `Overview of steps you follow to create a template`?

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+3. Create a new folder **test/fixtures** with a file named `asyncapi.yml` in your fixtures directory. This file is used to define the **structure** of your template. 
+4. Create a new file named `package.json` in your template directory. This file is used to define the **dependencies** for your template.
+5. Create a new file named `index.js` in your **template** directory. This file is used to define the **logic** for your template.
+6. Create a new directory named `partials` in your template directory. This directory is used to store the **partial HTML** files that are used to create the final documentation.
+7. Create a new directory named `assets` in your template directory. This directory is used to **store the CSS and JavaScript files that are used to style the final documentation or website**.
```
we do not cover that in this tutorial so we should not mention it, to not confuse

derberg replied:
more imporant is to mention `test/project/test.py` that you need to create a test

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+    "generator": ">=1.10.0 <2.0.0",
+    "supportedProtocols": ["mqtt"]
+  },
+  "dependencies": {
```
we are missing `devDependency` pointing to `rimraf`

```
    "devDependencies": {
        "rimraf": "^5.0.0"
    }
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+The code snippet above does the following:
+
+1. Import the `generator-react-sdk` dependency.
+2. The `asyncapi` argument is an instance of the [AsyncAPI Parser](https://www.asyncapi.com/docs/tools/generator/parser). It will allow you to access the content of the asyncapi document in your template using helper functions.
```
```suggestion
2. The `asyncapi` argument is an instance of the [AsyncAPI Parser](https://www.asyncapi.com/docs/tools/generator/parser). It will allow you to access the content of the AsyncAPI document in your template using helper functions.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+1. Create a new directory for your template named **python-mqtt-client-template**.
+2. Install the AsyncAPI CLI using the command `npm install -g @asyncapi/cli`.
+3. Create a new folder **test/fixtures** with a file named `asyncapi.yml` in your fixtures directory. This file is used to define the **structure** of your template. 
```
should we maybe refer to backgroud context section? like tell people that `asyncapi.yml` should contain content mentioned there?

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+  description: This service is in charge of processing all the events related to temperature.
+```
+
+`asyncapi.info().title()` will return Temperature Service.
````
```suggestion
`asyncapi.info().title()` returns `Temperature Service`.
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+Check out your shiny new generated files at output.
+```
+
+And navigating to the **project** directory, you should see a **client.py** file; the only content is Temperature Service.
````
```suggestion
And navigating to the **test/project** directory, you should see a **client.py** file; the only content is `Temperature Service`.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+Let's break down the previous command:
+
+- `asyncapi generate fromTemplate` is how you use AsyncAPI generator via the AsyncAPI CLI.
+- `asyncapi.yaml` points to your AsyncAPI document.
```
```suggestion
- `test/fixtures/asyncapi.yml` points to your AsyncAPI document.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+## Creating a template
+
+We will create an MQTT-supported template that will generate a Python client from the template and the AsyncAPI document above.
```
```suggestion
You will create an MQTT-supported template that will generate a Python client from the template and the AsyncAPI document above.
```

derberg replied:
best docs are when they are directed to reader. With blog posts it is different, but with docs it should be about the reader. Make sure you use it everywhere in the document

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+### Add parameters to the configuration file
+
+We often have different runtime environments in programming, e.g., development and production. You will use different servers to spin both of these instances. You'll probably have two broker versions, one for production and the other for development. You have defined a dev server in the AsyncAPI document:
```
```suggestion
You often have different runtime environments in programming, e.g., development and production. You will use different servers to spin both of these instances. You'll probably have two broker versions, one for production and the other for development. You have defined a dev server in the AsyncAPI document:
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+In this section, you'll:
+
+1. Write the MQTT client code.
+2. Write code to test the client works
+3. Update the template to use the client code.
+4. Setup a script to help you run this code
+5. Template your code.
```
```suggestion
1. Write the MQTT client code.
2. Write code to test the client works.
3. Update the template to use the client code.
4. Setup a script to help you run this code.
5. Template your code.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+1. Imports the MQTT module from the paho package, which provides the MQTT client functionality.
+2. Assigns the MQTT broker address `test.mosquitto.org` to the variable MQTT broker. This specifies the location where the MQTT client will connect to.
+3. Creates an instance of the MQTT client object. This object will be used to establish a connection with the MQTT broker and perform MQTT operations.
```
```suggestion
3. Defines an instance of the MQTT client object. This object will be used to establish a connection with the MQTT broker and perform MQTT operations.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+1. Imports the MQTT module from the paho package, which provides the MQTT client functionality.
+2. Assigns the MQTT broker address `test.mosquitto.org` to the variable MQTT broker. This specifies the location where the MQTT client will connect to.
+3. Creates an instance of the MQTT client object. This object will be used to establish a connection with the MQTT broker and perform MQTT operations.
+4. Establishes a connection to the MQTT broker. This command connects the MQTT client to the broker.
```
```suggestion
4. Defines that on client instance creation, it connects to the broker.
```

lets focus on code and not how someone will use it

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+2. Assigns the MQTT broker address `test.mosquitto.org` to the variable MQTT broker. This specifies the location where the MQTT client will connect to.
+3. Creates an instance of the MQTT client object. This object will be used to establish a connection with the MQTT broker and perform MQTT operations.
+4. Establishes a connection to the MQTT broker. This command connects the MQTT client to the broker.
+5. The MQTT topic to which the client will publish messages.
+6. This command publishes the temperature change information to the MQTT broker.
```
```suggestion
5. The `sendTemperatureChange` is a function that client user invokes to publish a message to the broker, and its specific topic.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+In summary, this code sets up an MQTT client using the paho-MQTT library. It connects to the `test.mosquitto.org` MQTT broker, and the `sendTemperatureChange()` method publishes temperature change information to the `temperature/changed` topic whenever called.
+
+### 2.Test the client
```
```suggestion
### 2. Test the client
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+
+```
+
+Run the code above on your terminal using the command `python test.py`. You'll should see output similar to the snippet below logged on your terminal:
````
```suggestion
Run the code above in your terminal using the command `python test.py`. You should see output similar to the snippet below logged on your terminal:
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+New temperature detected 72955029 sent to temperature/changed
+```
+
+To make sure your client works, also test it using an [MQTT CLI](https://hivemq.github.io/mqtt-cli/) using docker. Run the command `docker run hivemq/mqtt-cli sub -t temperature/changed -h test.mosquitto.org` on your teminal. It will download the image if you don't have it locally then the CLI will conncet to the broker, subscribe to the `temperature/changed` topic and then output the temperature ...
````
```suggestion
To make sure your `test.py` and client code works check if the broker really receives temperature-related messages. You can do it using an [MQTT CLI](https://hivemq.github.io/mqtt-cli/) using docker. Run the command `docker run hivemq/mqtt-cli sub -t temperature/changed -h test.mosquitto.org` in your terminal. It will download the image if you don't have it locally, then the CLI will connect to the broker, subscribe to the `temperature/changed` topic and then output the temperature ids on the terminal.
```

also please like in case of test.py terminal output, also show MQTT CLI client output

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+To make sure your client works, also test it using an [MQTT CLI](https://hivemq.github.io/mqtt-cli/) using docker. Run the command `docker run hivemq/mqtt-cli sub -t temperature/changed -h test.mosquitto.org` on your teminal. It will download the image if you don't have it locally then the CLI will conncet to the broker, subscribe to the `temperature/changed` topic and then output the temperature ...
+
+### 3.Update the template with client code
```
```suggestion
### 3. Update the template with client code
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+### 3.Update the template with client code
+
+Open **index.js** and copy the content of **client.py** and replace `{asyncapi.info().title()}` with it. It should look like the code snippet below now:
```
> copy the content of **client.py**

sorry? copy from where, it is not mentioned earlier in the document

derberg replied:
you are referring to `Create the client` and the code below sentence `The following code is an example of a Python client using the paho-MQTT library` right? but there is nothing about `client.py`

Florence-Njeri replied:
You're right. Thanks I'll update this

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+The 4 scripts above do the following:
+
+1. `test:clean`: This script uses the rimraf package to remove the old version of the file **test/project/client.py** everytime you run your test.
```
```suggestion
1. `test:clean`: This script uses the `rimraf` package to remove the old version of the file **test/project/client.py** every time you run your test.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+1. `test:clean`: This script uses the rimraf package to remove the old version of the file **test/project/client.py** everytime you run your test.
+2. `test:generate`: This script uses the AsyncAPI CLI to generate a new version of **client.py**
+3. `test:start`: This script runs the python code using **client.py**
+4. `test`: This script runs `npm test` to check that everything is working as expected.
```
```suggestion
4. `test`: This script runs all the other scripts in proper order.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+The 4 scripts above do the following:
+
+1. `test:clean`: This script uses the rimraf package to remove the old version of the file **test/project/client.py** everytime you run your test.
+2. `test:generate`: This script uses the AsyncAPI CLI to generate a new version of **client.py**
+3. `test:start`: This script runs the python code using **client.py**
```
```suggestion
2. `test:generate`: This script uses the AsyncAPI CLI to generate a new version of **client.py**.
3. `test:start`: This script runs the python code using **client.py**.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+servers:
+  dev:
+    url: test.mosquitto.org
+    protocol: mqtt
+    bindings:
+      mqtt:
+        clientId: temperature-service
```
```suggestion
    protocol: mqtt
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+servers:
+  dev:
+    url: test.mosquitto.org
+    protocol: mqtt
+    bindings:
+      mqtt:
+        clientId: temperature-service
```
```suggestion
    protocol: mqtt
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+Generator Error: This template requires the following missing params: server.
+```
+
+Update your `test:generate` script in `package.json` to include the server param `test:generate": "asyncapi generate fromTemplate test/fixtures/asyncapi.yml ./ --output test/project --force-write --param server=dev "`
````
```suggestion
Update your `test:generate` script in `package.json` to include the server param `test:generate": "asyncapi generate fromTemplate test/fixtures/asyncapi.yml ./ --output test/project --force-write --param server=dev"`
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+```
+
+1. Import the **Text** component that will wrap strings so they are indented properly in the output. Your import statement should now look like this: `import { File, Text } from '@asyncapi/generator-react-sdk'`.
+2. When the paho module import is rendered in **client.py** file, it will add two extra new lines.
````
please doublecheck with paho docs, how they spell it in docs, if lower case, I'm not sure but probably `Paho` or `PAHO`. But probably best is `Paho` always

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+1. Import the **Text** component that will wrap strings so they are indented properly in the output. Your import statement should now look like this: `import { File, Text } from '@asyncapi/generator-react-sdk'`.
+2. When the paho module import is rendered in **client.py** file, it will add two extra new lines.
+3. The broker url is templated in a Text component removing the `$` from the string template.
```
```suggestion
3. The broker url is templated in a `Text` component removing the `$` from the string template.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+1. Import the **Text** component that will wrap strings so they are indented properly in the output. Your import statement should now look like this: `import { File, Text } from '@asyncapi/generator-react-sdk'`.
+2. When the paho module import is rendered in **client.py** file, it will add two extra new lines.
+3. The broker url is templated in a Text component removing the `$` from the string template.
+4. Dynamically get the class name **TemperatureServiceClient** from the AsyncAPI document from the **info** object using the Parser API using the code: `asyncapi.info().title()` . It will return `Temperature Service`, then remove the spaces and add Client as a suffix.
```
```suggestion
4. Dynamically get the class name **TemperatureServiceClient** from the AsyncAPI document from the **info** object using the Parser API using the code: `asyncapi.info().title()` . It will return `Temperature Service`, then remove the spaces and add `Client` as a suffix.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+              self.client.connect(mqttBroker)
+
+  def sendTemperatureDrop(self, id):
+          topic = "temperature/droped"
```
```suggestion
          topic = "temperature/dropped"
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+
+```
+
+You'll then need to template to dynamically generate `sendTemperatureDropped` and `sendTemperatureRisen` functions in the generated code based off the AsyncAPI document content. The goal is to write template code that returns functions for channels that the Temperature Service application is subscribed to. The template code to generate these functions will look like this:
````
names of functions do not match the code

Florence-Njeri replied:
Nice catch

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+
+```js
+<Text newLines={2}>
+  <TopicVariable channels={asyncapi.channels().filterByReceive()} />
````
should be `TopicFunction` I think, not `TopicVariable`

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+}
+```
+
+`{ channels }`:the TopicFunction component accepts a custom prop called channels and in your template code.
````
```suggestion
`{ channels }`: the `TopicFunction` component accepts a custom prop called channels and in your template code
```

derberg commented on docs/generator_template.md:
````diff
@@ -0,0 +1,569 @@
+```
+
+`{ channels }`:the TopicFunction component accepts a custom prop called channels and in your template code.
+`getTopics(channels)`: Returns a list of objects, one for each channel with two properties; name and topic. The **name** holds information about the operationId provided in the AsyncAPI document while the **topic** holds information about the address of the topic.
````
```suggestion
`getTopics(channels)`: Returns a list of objects, one for each channel with two properties; name and topic. The **name** holds information about the `operationId` provided in the AsyncAPI document while the **topic** holds information about the address of the topic.
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+`{ channels }`:the TopicFunction component accepts a custom prop called channels and in your template code.
+`getTopics(channels)`: Returns a list of objects, one for each channel with two properties; name and topic. The **name** holds information about the operationId provided in the AsyncAPI document while the **topic** holds information about the address of the topic.
+
+Import the `TopicFunction` component in your template code in **index.js** and add the template code to generate the functions to topics that the Temperature Service application is subscribed to. In your case, .The final version of your template code should look like this:
```
```suggestion
Import the `TopicFunction` component in your template code in **index.js** and add the template code to generate the functions to topics that the `Temperature Service` application is subscribed to. In your case, the final version of your template code should look like this:
```

derberg reviewed (changes requested):
Nice work 👏🏼 left some different style/technical review comments

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,569 @@
+
+In this tutorial, you'll learn to create a simple generator template using a Python MQTT client. You'll use the AsyncAPI document and the template you develop to generate Python code. Additionally, you'll create template code with a reusable component to reuse the custom functionality you create and test your code using an MQTT client.
+
+Suppose you can only sleep when the AC in your bedroom is set to 22 °C, and you can't see when the temperature drops or rises above that. You can install a smart monitor in your bedroom that keeps track of the temperature and notifies you to adjust it to your optimum temperature if it fluctuates. You will create a template that alerts you when the temperature fluctuates from 22 °C.
```
btw, love this idea, I felt like a motivation kick when I read it, to continue reading more. Great story based tutorial writing 💪🏼

Florence-Njeri replied:
Thank you @derberg

Florence-Njeri commented:
@thulieblack @derberg @pratik2315 I've updated this PR with all the awesome feedback and suggestions you'd left me. On top of that, I made a few more editorial reviews to fix typos, grammar, and bold file names instead of showing them as `code`. I also wanted to get your input on: 
1. Is the title too long?
![](https://imgur.com/42K5f4O.png)
2. What do you think of the outline?
![](https://imgur.com/7NsekQs.png)
3. I renamed the conclusion section to where to go from here let me know if this makes sense or if I should stick to conclusion
![](https://imgur.com/pi2xX5d.png)
4. Should I add Alejandra as a reviewer to this PR too?

thulieblack commented:
yo @Florence-Njeri can you share the preview link here as well

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+
+</Remember>
+
+## Overview of steps to follow to create a template
```
regarding screenshot of outline, his section looks super long there

```suggestion
## Overview of steps
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+
+Great job completing this tutorial! You have learnt how to use an AsyncAPI file to create a Python MQTT template and used it with the Paho-MQTT library in Python to connect to an MQTT broker and publish messages.:]
+
+If you want to tinker with a completed template and see what it would look like in production, check out the [Paho-MQTT template](https://github.com/derberg/python-mqtt-client-template) developed by [Lukasz](https://github.com/derberg). You can also check out the accompanying [tutorial to create MQTT client code](https://derberg-github-io-git-articlecodegenpython-derberg.vercel.app/blog/asyncapi- ...
```
```suggestion
If you want to tinker with a completed template and see what it would look like in production, check out the [Paho-MQTT template](https://github.com/derberg/python-mqtt-client-template/tree/v1.0.0). You can also check out the accompanying [article about creating MQTT client code](https://www.brainfart.dev/blog/asyncapi-codegen-python).
```

derberg reviewed (changes requested):
left 2 comments, one related to table of contents outline

regarding title of the document, yeah - long, but what can you do really. We can only be consistent and not calling it with `How to` as other tutorials do not have it. How about `Creating a simple generator template`?still will be long I guess


@thulieblack there is no preview link, @Florence-Njeri is for sure building it manually on local

@Florence-Njeri important to notice is that whatever is located in `generator` repo lands under `asyncapi.com/tools/generator` and not in `Tutorials` bucket. This tutorial is super specific to the Generator and its functionality

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+weight: 170
+---
+
+In this tutorial, you'll learn to create a simple generator template using a Python MQTT client. You'll use the AsyncAPI document and the template you develop to generate Python code. Additionally, you'll create template code with a reusable component to reuse the custom functionality you create and test your code using an MQTT client.
```
```suggestion
This tutorial teaches you how to create a simple generator template using a Python MQTT client. You'll use the AsyncAPI document and the template you develop to generate Python code. Additionally, you'll create template code with a reusable component to reuse the custom functionality you create and test your code using an MQTT client.
```

thulieblack commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+
+In this tutorial, you'll learn to create a simple generator template using a Python MQTT client. You'll use the AsyncAPI document and the template you develop to generate Python code. Additionally, you'll create template code with a reusable component to reuse the custom functionality you create and test your code using an MQTT client.
+
+Suppose you can only sleep when the AC in your bedroom is set to 22 °C, and you can't sleep when the temperature drops or rises above that. You can install a smart monitor in your bedroom that keeps track of the temperature and notifies you to adjust it to your optimum temperature when it fluctuates. You will create a template to alerts you when the bedroom's temperature fluctuates from 22 °C.
```
```suggestion
Suppose you can only sleep when the AC in your bedroom is set to 22 °C, and you can't sleep when the temperature drops or rises above that. You can install a smart monitor in your bedroom that keeps track of the temperature and notifies you to adjust it to your optimum temperature when it fluctuates. You will create a template to alert you when the bedroom's temperature fluctuates from 22 °C.
```

thulieblack commented on docs/generator_template.md:
````diff
@@ -0,0 +1,575 @@
+        self.client.publish(topic, id)
+```
+
+Make sure you have the Paho-mqtt library installed. You can install it using pip with the `pip install paho-mqtt` command.
````
```suggestion
Make sure you have the Paho-MQTT library installed. You can install it using pip with the `pip install paho-mqtt` command.
```

thulieblack commented:
1. How about we simplify the title to **Generator Template**, since the first paragraph already explains that we are creating a simple generator template
>  In this tutorial, you'll learn to create a simple generator template using a Python MQTT client.

derberg reviewed (commented):
@thulieblack 

> How about we simplify the title to Generator Template, since the first paragraph already explains that we are creating a simple generator template

you mean title of the doc? tutorials should have a title that says what it is about

`Generator Template` says it is about generator template
`Creating a simple generator template` says it is about creating a simple generator

We also should be consistent with other titles of the docs on the website.

thulieblack commented:
Mmmh I see, then I guess we can go with `Creating a Generator Template`

Florence-Njeri commented:
> yo @Florence-Njeri can you share the preview link here as well

Hey @thulieblack I was running it in localhost. How do I get a preview link 🤔

Florence-Njeri commented:
> sounds good to me, as also word `simple` is no good for titles
> 
> @Florence-Njeri ready for final round?

Yes it is

Florence-Njeri commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+---
+title: "How to create a simple generator template"
```
```suggestion
title: "Creating a generator template"
```

derberg commented:
> Hey @thulieblack I was running it in localhost. How do I get a preview link

there is no quick way unfortunately. The only one I know for now is that you use GitPod to get a virtual env for website:
1. go here http://gitpod.io/#https://github.com/asyncapi/website
2. manually put new markdown in a desired location
3. stop server and run there again `npm run dev` 
4. in logs you see localhost link, click it, and it will get you to a separate website like https://3000-asyncapi-website-q3b1xuns0yu.ws-eu101.gitpod.io/docs/tools/generator/generator-template

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+---
+title: "Creating a generator template"
```
```suggestion
title: "Creating a Generator template"
```

derberg commented on docs/generator_template.md:
```diff
@@ -0,0 +1,575 @@
+
+## Where to go from here?
+
+Great job completing this tutorial! You have learnt how to use an AsyncAPI file to create a Python MQTT template and used it with the Paho-MQTT library in Python to connect to an MQTT broker and publish messages.:]
```
> :]

is this one intended? you can use real emoji if you want

derberg reviewed (commented):
just 2 small comments

and regarding navigation, maybe under `Template your code` section, all subsections should also have number, so in navigation it then makes more sense? something like:

![Screenshot 2023-07-19 at 10 03 22](https://github.com/asyncapi/generator/assets/6995927/71cd4df2-d226-45f9-a62f-305fe5e43328)

derberg reviewed (approved):
LGTM

@thulieblack anything from your side?

thulieblack reviewed (approved):
Looks great from my side, well done @Florence-Njeri 
![BlondeHairCuteDanceGIF](https://github.com/asyncapi/generator/assets/66913810/4ffd19f8-d073-4d06-b161-c280a5e782ab)

derberg commented:
We need @pratik2315 approval but he is completely unreachable, tried in the past slack - email - twitter

just sent invite to LinkedIn

if we will not respond I will:
- merge as admin, ignoring his lack of approval
- and we need to remove him as a maintainer

derberg commented:
@Florence-Njeri congrats for the amazing work 👏🏼  https://www.asyncapi.com/docs/tools/generator/generator_template (we are just a bit inconsistent as filename is `generator_template` and all other have `-`) and now when I'm thinking, since it is `generator` docs, the document title could probably be just `Creating a template` 😄 
@thulieblack thanks a lot for help with review, I guess you can share with @Florence-Njeri how to get the bounty

when you feel in the mood, please start an issue in `website` repo to discuss if there is a need for content buckets for tools long term, as we have this use case where we have lots of docs, but as flat list:
- we have no way to visualize that something is a tutorial
- but we have a use case that it is very tool specific/oriented tutorial and it cannot be in the main "tutorials" bucket

thulieblack commented:
Congratulations @Florence-Njeri, on successfully finishing the task🎉🎉. Please submit an expense on [open collective](https://opencollective.com/asyncapi) to claim your bounty

Florence-Njeri commented:
> @Florence-Njeri congrats for the amazing work 👏🏼 https://www.asyncapi.com/docs/tools/generator/generator_template (we are just a bit inconsistent as filename is `generator_template` and all other have `-`) and now when I'm thinking, since it is `generator` docs, the document title could probably be just `Creating a template` 😄 @thulieblack thanks a lot for help with review, I guess you can share with @Florence-Njeri how to get the bounty
> 
> when you feel in the mood, please start an issue in `website` repo to discuss if there is a need for content buckets for tools long term, as we have this use case where we have lots of docs, but as flat list:
> 
> * we have no way to visualize that something is a tutorial
> * but we have a use case that it is very tool specific/oriented tutorial and it cannot be in the main "tutorials" bucket

I'll create an issue and put up a follow up PR for this

aeworxet commented:
/bounty
